src/engine/IndexScan.cpp

```diff
@@ -134,40 +185,74 @@ VariableToColumnMap IndexScan::computeVariableToColumnMap() const {
   return variableToColumnMap;
 }
 
+//______________________________________________________________________________
+std::shared_ptr<QueryExecutionTree> IndexScan::makeCopyWithAddedPrefilters(
+    PrefilterIndexPair prefilter) {
+  return ad_utility::makeExecutionTree<IndexScan>(
+      getExecutionContext(), permutation_, subject_, predicate_, object_,
+      additionalColumns_, additionalVariables_, std::move(graphsToFilter_),
+      std::move(prefilter));
+}
+
+//______________________________________________________________________________
+std::vector<CompressedBlockMetadata> IndexScan::applyFilterBlockMetadata(
+    const std::vector<CompressedBlockMetadata> blocks) const {
+  if (prefilter_.has_value()) {
+    auto& prefilterIndexPair = prefilter_.value();
+    return prefilterExpressions::detail::evaluatePrefilterExpressionOnMetadata(
+        prefilterIndexPair.first->clone(), blocks, prefilterIndexPair.second);
+  }
+  return blocks;
+};
+
 // _____________________________________________________________________________
-Result::Generator IndexScan::scanInChunks() const {
+Result::Generator IndexScan::chunkedIndexScan() const {
   auto metadata = getMetadataForScan();
   if (!metadata.has_value()) {
     co_return;
   }
   auto blocksSpan =
       CompressedRelationReader::getBlocksFromMetadata(metadata.value());
-  std::vector<CompressedBlockMetadata> blocks{blocksSpan.begin(),
-                                              blocksSpan.end()};
-  for (IdTable& idTable : getLazyScan(std::move(blocks))) {
+  for (IdTable& idTable : getLazyScan({blocksSpan.begin(), blocksSpan.end()})) {
     co_yield {std::move(idTable), LocalVocab{}};
   }
 }
 
 // _____________________________________________________________________________
-ProtoResult IndexScan::computeResult(bool requestLaziness) {
-  LOG(DEBUG) << "IndexScan result computation...\n";
-  if (requestLaziness) {
-    return {scanInChunks(), resultSortedOn()};
-  }
-  IdTable idTable{getExecutionContext()->getAllocator()};
-
+IdTable IndexScan::materializedIndexScan() const {
+  // Get the blocks.
+  auto metadata = getMetadataForScan();
+  auto blockSpan =
+      metadata.has_value()
+          ? CompressedRelationReader::getBlocksFromMetadata(metadata.value())
+          : std::span<const CompressedBlockMetadata>{};
+  auto optFilteredBlocks = getLimit().isUnconstrained()
+                               ? std::optional{applyFilterBlockMetadata(
+                                     {blockSpan.begin(), blockSpan.end()})}
+                               : std::nullopt;
```

Comment:
Factor this out into a function
`optional<pair<Metadata, BlockVector>>  getMetadataAndPossiblyPrefilteredBlocks`
and use this in both the `chunkedIndexScan` and the `materializedIndexScan`.

Reply:
Or better:
two functions
1. metadata and complete blocks
2. filter blocks if no metadata (the thing with the `isUnconstrained`.